File: 2/js/src/ThreeNumSum.ts
function solve(list: number[], target: number): (number[])[] {
	const n = list.length;

	list.sort((a, b) => a-b);

	let res: (number[])[] = [];
	for(let i=0; i<list.length-2; i++) {
		const x = list[i];

		let left = i+1;
		let right = n-1;

		while(left < right) {
			const y = list[left];
			const z = list[right];

			if(x + y + z < target) {
				left++;
			} else if(x + y + z > target) {
				right--;
			} else {
				res.push([x,y,z,]);
				left++;
				right--;
			}
		}
	}

	return res;
}

(function main() {
	const list = [12,3,1,2,-6,5,-8,6];
	const target = 0;

	console.log(solve(list, target));
})();

File: 2/java/ThreeNumSum.java
import java.util.Arrays;
import java.util.ArrayList;

class ThreeNumSum {
	
	public static ArrayList<int[]> threeNumSum(int array[], int target) {
		ArrayList pairs = new ArrayList<int[]>();	

		Arrays.sort(array);
		int n = array.length;	
		
		for(int i=0; i<=n-3; i++) {
			int left_ptr = i+1;
			int right_ptr = n-1;
			
			while(left_ptr < right_ptr) {
				if(array[left_ptr] + array[right_ptr] + array[i] == target) {
					pairs.add(new int[]{array[i], array[left_ptr], array[right_ptr]});
					left_ptr++;
					right_ptr--;
				} else if(array[left_ptr] + array[right_ptr] + array[i] < target) {
						left_ptr++;	
				} else if(array[left_ptr] + array[right_ptr] + array[i] > target) {
						right_ptr--;
				}
			}

		}
		
		return pairs;
	}

	public static void main(String args[]) {
		int array[] = new int[]{12,3,1,2,-6,5,-8,6};
		int target = 0;

		ArrayList pairs = threeNumSum(array, target);

		for(int i=0; i<pairs.size(); i++) {
			int pair[] = (int[])pairs.get(i);
			for(int number : pair) {
				System.out.print(String.format("%d \t", number));
			}
			System.out.println();
		}
	}
}
